add support for cpuset.mems

hqhq opened the pull request.

SvenDowideit commented on docs/man/docker-create.1.md:
```diff
@@ -60,9 +61,12 @@ IMAGE [COMMAND] [ARG...]
    CPUs in which to allow execution (0-3, 0,1)
 
+**--cpuset-mems**=""
+   MEMs in which to allow execution (0-3, 0,1)
```
mmm, i think you will need to tell the reader what an `MEMs` is - CPU possibly not - @fredlf ?

hqhq replied:
ok, I'll add that, thanks.

fredlf replied:
Nah, I don't think we need to define CPU. But yes, we need to define MEMs.

SvenDowideit commented:
@hqhq in run.md, we should also give the user an example, and explain what it does.

LK4D4 commented:
@hqhq Docker API still used by users.

hqhq commented:
@LK4D4 Hmmm.... what's your suggestion then? Should we keep using Cpuset
and just add CpusetMems? 
Or do we have some deprecated tags for API?

LK4D4 commented:
@hqhq I don't know actually :) There is way to check API version, but I'm not sure how this can help. Only way I see - deprecation in docs :) Maybe @crosbymichael knows better way

hqhq commented:
@crosbymichael @LK4D4 what do you think if we go this way:
1. add CpusetCpus which has the same function as Cpuset, and keep using Cpuset;
2. add deprecated notify mechanism for docker http APIs, and mark Cpuset as deprecated;
3. remove Cpuset after a period of time;

LK4D4 commented:
@hqhq I am okay with this. But we're not very good with removing things :) Maybe in 2.0.

fredlf commented:
Agree with @SvenDowideit  that we need an example in run.md and should define MEMs.

hqhq commented:
@SvenDowideit @fredlf @LK4D4 @crosbymichael 
Sorry for the delay, rebased and updated, please review again, thanks.

LK4D4 commented on docs/sources/reference/api/docker_remote_api_v1.12.md:
```diff
@@ -115,7 +115,8 @@ Create a container
              "MemorySwap":0,
              "CpuShares": 512,
-             "Cpuset": "0,1",
+             "CpusetCpus": "0,1",
```
Not sure that it is good idea to change docs for previous versions. ping @SvenDowideit

hqhq replied:
You are right, I think we should not change docs for previous version.
For now, change v1.16 and v1.17 is good enought.

hqhq commented:
updated, remove the change of previous version of docs.

LK4D4 commented:
@hqhq I'm not sure about 1.16 too, seems like this version was already released.

crosbymichael commented:
We can look at removing it in the future but I think the FIXME is enough for now as along as we keep supporting Cpuset as the same as CpusetCpus

hqhq commented:
@LK4D4 Yeah, right, sorry about that. 
Updated.

SvenDowideit commented:
Ah, thats a little more useful, but an example :)

hqhq commented:
@SvenDowideit do you mean we should add some more example in doc to show
how to use CpusetMems?

vbatts commented:
ping @crosbymichael

icecrime commented on pkg/tarsum/tarsum_test.go:
```diff
@@ -49,13 +49,13 @@ var testLayers = []testLayer{
 		jsonfile: "testdata/xattr/json",
 		version:  Version0,
-		tarsum:   "tarsum+sha256:e86f81a4d552f13039b1396ed03ca968ea9717581f9577ef1876ea6ff9b38c98"},
+		tarsum:   "tarsum+sha256:ad39b3cb1cc73565c8afb5add3e8126383dff55c6dda6f8e67665ca81f4fb839"},
```
This seems like a completely unrelated diff: could you please revert?

icecrime commented on pkg/tarsum/tarsum_test.go:
```diff
@@ -49,13 +49,13 @@ var testLayers = []testLayer{
 		jsonfile: "testdata/xattr/json",
 		version:  VersionDev,
-		tarsum:   "tarsum.dev+sha256:6235cd3a2afb7501bac541772a3d61a3634e95bc90bb39a4676e2cb98d08390d"},
+		tarsum:   "tarsum.dev+sha256:4eb104c8272535d05c06dd3d8e48aceb4eae0b9206c8151410cfd20839238ac8"},
```
This seems like a completely unrelated diff: could you please revert?

icecrime replied:
If the new fields were not part of the image, updating those tarsums wouldn't be necessary.

icecrime commented on pkg/tarsum/testdata/xattr/json:
```diff
@@ -1 +1 @@
-{"id":"4439c3c7f847954100b42b267e7e5529cac1d6934db082f65795c5ca2e594d93","parent":"73b164f4437db87e96e90083c73a6592f549646ae2ec00ed33c6b9b49a5c4470","created":"2014-05-16T17:19:44.091534414Z","container":"5f92fb06cc58f357f0cde41394e2bbbb664e663974b2ac1693ab07b7a306749b","container_config":{"Hostname":"9565c6517a0e","Domainname":"","User":"","Memory":0,"MemorySwap":0,"CpuShares":0,"Cpuset":"","Att ...
\ No newline at end of file
+{"id":"4439c3c7f847954100b42b267e7e5529cac1d6934db082f65795c5ca2e594d93","parent":"73b164f4437db87e96e90083c73a6592f549646ae2ec00ed33c6b9b49a5c4470","created":"2014-05-16T17:19:44.091534414Z","container":"5f92fb06cc58f357f0cde41394e2bbbb664e663974b2ac1693ab07b7a306749b","container_config":{"Hostname":"9565c6517a0e","Domainname":"","User":"","Memory":0,"MemorySwap":0,"CpuShares":0,"CpusetCpus":"", ...
```
This seems like a completely unrelated diff: could you please revert?

icecrime replied:
Same here: the new fields probably shouldn't appear here.

hqhq replied:
I'm not sure what is this tarsum for, but if it contains config information, why
shouldn't the new fields in config appear there?

icecrime replied:
The relevant comment is found in [`runconfig/config.go:8`](https://github.com/docker/docker/pull/9536/files#diff-2660508eef7a8eeadb6d27b8bbf6d718R10):

> // Note: the Config structure should hold only portable information about the container.
> // Here, "portable" means "independent from the host we are running on".
> // Non-portable information *should* appear in HostConfig.

The `Cpuset` field is already in there, but that was a mistake: we're stuck with it for backward compatibility. However, the new fields `CpusetCpus` and `CpusetMems` don't have to be added to the `Config` struct, which will have the nice side-effect of not appearing in the image JSON, which doesn't require you to update the tarsums.

hqhq replied:
ok, I get it, so all cgroups related fields will be added to hostconfig then.

icecrime replied:
Thank you :+1:

icecrime commented on runconfig/merge.go:
```diff
@@ -20,6 +20,17 @@ func Merge(userConf, imageConf *Config) error {
+		userConf.Cpuset = imageConf.Cpuset
+		userConf.CpusetCpus = imageConf.CpusetCpus
+	}
+
```
This is unnecessary: we don't want these fields to appear in the image (it's host dependent), so the merging behavior shouldn't be required.

hqhq replied:
Can you explain more why we don't want these fields to appear in the image?
From what I understanding, this part is for `POST /commit`, users can send a json
file in the request which might contain the old cpuset entry, that was supposed to
be added in the image, right? 
And users can inspect an image through `GET /images/(name)/json`, they might
want to get their old `Cpuset` entry.

SvenDowideit commented:
@hqhq yes, an example that gives a micro explaination of how to use the option, what the results are and importantly, why you'd want to do it.

hqhq committed:
cgroup: add test case for cpuset.mems

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

fredlf commented:
Agree with @SvenDowideit , this needs a simple usage example.

fredlf commented on docs/man/docker-create.1.md:
```diff
@@ -61,9 +62,13 @@ IMAGE [COMMAND] [ARG...]
    CPUs in which to allow execution (0-3, 0,1)
 
+**--cpuset-mems**=""
+   MEMs in which to allow execution (0-3, 0,1), MEMs means memory nodes,
```
Please simplify and remove run-on: "Memory nodes (MEMs) in which to allow execution (0-3, 0, 1). Only effective on NUMA systems."

hqhq replied:
Will do, thanks for the suggestion.

fredlf commented on docs/man/docker-run.1.md:
```diff
@@ -94,9 +95,13 @@ run**.
 
+**--cpuset-mems**=""
+   MEMs in which to allow execution (0-3, 0,1), MEMs means memory nodes,
+   this is only effective on NUMA system.
```
As above.

fredlf commented on docs/sources/reference/api/docker_remote_api_v1.17.md:
```diff
@@ -178,7 +179,10 @@ Json Parameters:
+-   **CpuSet** - The same as CpusetCpus, but obsoleted, please don't use.
+-   **CpuSetCpus** - String value containg the cgroups CpusetCpus to use.
+-   **CpuSetMems** - String value containg the cgroups CpusetMems to use.
+      Mems means memory nodes, only effective on NUMA system.
```
As above.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -723,7 +723,9 @@ Creates a new container.
-      --cpuset=""                CPUs in which to allow execution (0-3, 0,1)
+      --cpuset-cpus=""           CPUs in which to allow execution (0-3, 0,1)
+      --cpuset-mems=""           MEMs in which to allow execution (0-3, 0,1)
+                                   MEMs means memory nodes, only effective on NUMA system.
```
As above.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1550,7 +1552,9 @@ removed before the image is removed.
-      --cpuset=""                CPUs in which to allow execution (0-3, 0,1)
+      --cpuset-cpus=""           CPUs in which to allow execution (0-3, 0,1)
+      --cpuset-mems=""           MEMs in which to allow execution (0-3, 0,1)
+                                   MEMs means memory nodes, only effective on NUMA system.
```
As above.

hqhq committed:
add support for cpuset.mems

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
updated.
 - move CpusetCpus and CpusetMems from Config to hostConfig
 - remove some unnecessary compatible code
 - remove changes for tarsum_test
 - updated comments for CpusetMems, added an example

LK4D4 commented:
@hqhq Tests are failing - see drone.

hqhq commented:
Sorry, fixed the test case failing.

hqhq commented:
I don't know why it's still failing, error message is:
panic: test timed out after 30m0s

why it's taking so long? Can anyone help to take a look?
Thanks.

hqhq commented:
Seems like this error happens sometimes and now drone is passed.

Guys, can you help review again? This PR has been taken too long, and blocking my
TODO list. :(

LK4D4 commented:
LGTM
tests not very impressive, but I think we can improve that in separate PR
ping @crosbymichael

hqhq commented:
@LK4D4 breaking what exactly?

LK4D4 commented:
`config.Cpuset` is not used anymore, so if someone was relying on this feature in api - we will break his  workflow.

hqhq commented:
No it won't break that, `config.Cpuset` is still used, you can see the changelog
in my first patch, that's how compatibility works for now.

LK4D4 commented:
@hqhq sorry if I doesn't get, but here:
https://github.com/hqhq/docker/commit/bfea793d0af4c2a489aa6fdcc886e25aa312a738#diff-6ce9e79ddb91a3f06352abe1f2c72ecbR280
I see that now `config.Cpuset` is not passed to `execdriver`. So, if I understand right then `Cpuset` is still present in `config` but not used for container setup. Maybe I just don't understand what's going on :) ping @crosbymichael

hqhq commented:
Yes, it is not passed to `execdriver`, `config.Cpuset` is only set for showing
users the config info, because the value is the same as `hostconfig.CpusetCpus`,
so in other parts of docker code, we use `hostconfig.CpusetCpus` instead.

One exception is that when users use `CpusetCpus` in jaon for their http APIs,
we won't setup `config.Cpuset` because the user won't expect that.

LK4D4 commented:
@hqhq I'm talking about case when use use `Cpuset` in json. There is no code which setups `CpusetCpus` from this, it was before in merge, but now I don't see it :/

hqhq committed:
rename Cpuset to CpusetCpus

As we will add CpusetMems, CpusetCpus is definitely a better name, but some
users are already using Cpuset in their http APIs, so we need time to change
it, right now we just make it compatible.

A significant change is that we add CpusetCpus in hostConfig, not Config,
because it's host dependent.

For compatibility, the main idea is keep using Cpuset in Config Struct, and
make it has the same value as CpusetCpus, but not always, some scenarios:
 - Users use --cpuset in docker command, it can setup cpuset.cpus and can
   get Cpuset field from docker inspect or other http API which will get
   config info.
 - Users use --cpuset-cpus in docker command, ditto.
 - Users use Cpuset field in their http APIs, ditto.
 - Users use CpusetCpus field in their http APIs, they won't get Cpuset field
   in Config info, because by then, they should already know what happens
   to Cpuset.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@LK4D4 Yep, that's the point I've missed in this version, sorry about that,
since we don't put CpusetCpus in config, so merge function is no longer 
the right place, I put the sync in `ContainerHostConfigFromJob`, when 
container start, `CpusetCpus` will get the right value if people still use `Cpuset`.

Update:
 - add sync for `CpusetCpus` and `Cpuset`, when people use `Cpuset` in json,
   it will still be workable.

LK4D4 commented:
Now seems legit :)

unclejack commented:
LGTM

crosbymichael commented:
@LK4D4 @unclejack shouldn't we wait until we can move all the resource config options to the hostconfig instead of having 6 on Config and 2 on hostconfig?

unclejack commented:
@crosbymichael Fair enough, I'll remove my approval.

crosbymichael commented:
This is something that we will have to do because @hqhq has already helped alot and has been awesome on expanding cgroup support for docker.  

@hqhq sorry for the confusion with this PR.  Cgroup resource options have been on the `Config` since forever.  They are host specific options so being on a config makes images not portable as this can be committed to an image.  HostConfig is the correct place for this but I just want to make sure when we make the switch we don't have any regressions and we move everything so that it's not confusing to the user, as these structs are also used in the API.

Let us discuss a little more about a game plan for these resource settings and we can do the work if any more changes are required.  

Thanks again!!

hqhq commented:
@crosbymichael agreed, we should stop all the confusion of cgroup resource options, but this might be done step by step, and I think this PR is a good place to start.

What we need to do now is to confirm a proper way to do the compatibility work, don't have any regressions and make sure people can know how to use it the right way. So we can copy it to all other options. And I'd like to do this work :)

What do you think? BTW, do you think the compatibility work in this PR is good enough for all other options?

crosbymichael commented:
@hqhq that would be awesome if you are interested in doing this work.  

I think at the API request level we should do the copy from `Config` to `HostConfig` for compatibility then null out the fields on `Config`.

Then if you just move all the cgroup resource options to hostconfig and make sure it's copied at the API level it would be good to go.  

Thanks for your help on this!

SvenDowideit commented:
TBH, the text of the docs is now sufficiently detailed, that an example isn't going to be useful - great job! If we were to add an example, it would draw more attention to it from readers that do _not_ have a NUMA system, and we should not make them (aka the majority) jealous :)

Docs LGTM - @fredlf @jamtur01

jamtur01 commented:
Should there be something in the run.md docs? Otherwise LGTM.

hqhq commented:
Thanks for all your reviews, let's wait for PR https://github.com/docker/docker/pull/10298 right now.

fredlf commented on docs/man/docker-create.1.md:
```diff
@@ -63,9 +64,16 @@ IMAGE [COMMAND] [ARG...]
+   Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
+
+   If you have 4 memory nodes on your system(0-3), use `--cpuset-mems=0,1`
+then process in your docker container will only use memory from the first
```
processes, Docker

fredlf commented on docs/man/docker-run.1.md:
```diff
@@ -96,9 +97,15 @@ run**.
+**--cpuset-mems**=""
+   Memory nodes (MEMs) in which to allow execution (0-3, 0,1). Only effective on NUMA systems.
+   If you have 4 memory nodes on your system(0-3), use `--cpuset-mems=0,1`
+then process in your docker container will only use memory from the first
```
As above.

fredlf commented:
Copy edits, otherwise LGTM. Agree with Sven, example no longer really needed.

LK4D4 commented:
@hqhq This needs rebase.

crosbymichael commented:
We will close this one for now in favor of the work being done in #10298 to allow this change to be merged.

unclejack commented:
@crosbymichael Can we reopen this PR now? #10298 has been merged.

hqhq commented:
@unclejack codes are quite different, I just send a new PR to support cpuset.mems, https://github.com/docker/docker/pull/12139, please take a look.
